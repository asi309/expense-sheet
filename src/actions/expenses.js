import database from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (
    { 
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
) => {
    return (dispatch) => {
        const expense = { description, note, amount, createdAt }
        return database.ref('expenses').push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (id) => {
    return (dispatch) => {
        database.ref(`expenses/${id}`).remove()
        .then(() => {
            dispatch(removeExpense(id));
        });
    };
};

export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        database.ref(`expenses/${id}`).update(updates)
        .then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// export const clearState = () => ({
//     type: "CLEAR_STATE"
// });

// export const startSetExpenses = () => {
//     return (dispatch) => {
//         database.ref('expenses').on('value', (snapshot) => {
//             dispatch(clearState());
//             const val = snapshot.val();
//             for(const key in val){
//                 dispatch(addExpense({
//                     id: key,
//                     ...val[key]
//                 }));
//             }
//         });
//     };
// }