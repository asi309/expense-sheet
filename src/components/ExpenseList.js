import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getFilteredExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.list === 0 ? (
                <p>No expenses</p>
            ):(
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={ expense.id } { ...expense } />;
                })
            )
        }
    </div>
);

const mapStateToProps = (state) => ({
    expenses: getFilteredExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
