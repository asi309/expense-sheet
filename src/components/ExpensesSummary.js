import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getFilteredExpense from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
    <div>
        {
            props.expensesCount === 1 &&
            <h4>Viewing 1 expense totalling {numeral(props.totalExpenses / 100).format('$0,0.00')}</h4>
        }
        {
            props.expensesCount > 1 &&
            <h4>Viewing {props.expensesCount} expenses totalling
                {numeral(props.totalExpenses / 100).format('$0,0.00')}
            </h4>
        }
    </div>
);

const mapStateToProps = (state) => ({
    expensesCount: getFilteredExpense(state.expenses, state.filters).length,
    totalExpenses: getExpensesTotal(state.expenses)
});

export default connect(mapStateToProps)(ExpensesSummary);