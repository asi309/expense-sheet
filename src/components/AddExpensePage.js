import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export const AddExpensePage = (props) => (
    <div>
        <h3>Add Expense</h3>
        <ExpenseForm 
            onSubmit = {(expense) => {
                props.onSubmit(expense);
                props.history.push('/');
            }}
        />
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);