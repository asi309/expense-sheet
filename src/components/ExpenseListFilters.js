import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends Component {
    state = {
        focused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (focused) => {
        this.setState(() => ({ focused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortByChange = (e) => {
        if(e.target.value === 'date'){
            this.props.setSortByDate();
        }else{
            this.props.setSortByAmount();
        }
    };

    render(){
        return (
            <div>
                <input
                    type='text'
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={ this.props.filters.startDate }
                    startDateId = "start-date"
                    endDate={ this.props.filters.endDate }
                    endDateId = "end-date"
                    onDatesChange={ this.onDatesChange }
                    focusedInput={ this.state.focused }
                    onFocusChange={ this.onFocusChange }
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setSortByAmount: () => dispatch(sortByAmount()),
    setSortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);