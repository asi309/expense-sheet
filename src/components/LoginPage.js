import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div>
        {
        //     <form>
        //     <input
        //         type="text"
        //         placeholder="username or email"
        //         autoFocus
        //     />
        //     <input
        //         type="password"
        //         placeholder="password"
        //     />
        //     <button onClick={ startLogin }>Submit</button>
        // </form>
    }
        <button onClick={ startLogin }>Login</button>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);