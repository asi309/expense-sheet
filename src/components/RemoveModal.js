import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import { startRemoveExpense } from '../actions/expenses';

export const RemoveModal = (props) => {
    Modal.setAppElement(document.getElementById('app'));
    return (
        <Modal
            isOpen = {props.isOpen}
            handleModalClose = {props.handleModalClose}
            contentLabel = "Remove Expense"
            closeTimeoutMS = {250}
            className="modal"
        >
            <h3 className="modal__title">Remove Expense</h3>
            <p className="modal__body">Are you sure you want to remove this expense?</p>
            <button
                className="button button--warning"
                onClick={props.handleRemoveExpense}
            >
                Remove
            </button>
            <button
                className="button"
                onClick={props.handleModalClose}
            >
                Cancel
            </button>
        </Modal>
    );
};

const MapDispatchToProps = (dispatch) => ({
    onRemove: (id) => dispatch(startRemoveExpense(id))
});

export default connect(undefined, MapDispatchToProps)(RemoveModal);