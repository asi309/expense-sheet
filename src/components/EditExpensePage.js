import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import RemoveModal from './RemoveModal';

export const EditExpensePage = (props) => {
    const [modalIsOpen, setIsOpen] = React.useState(false);
    const openModal = () => {
        setIsOpen(true);
    };
    const closeModal = () => {
        setIsOpen(false);
    };
    const handleRemoveExpense = (e) => {
        props.onRemove(props.expense.id);
        props.history.push('/');
    };
    
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Edit Expense</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm
                    expense={props.expense}
                    onSubmit={(expense) => {
                        props.onSubmit(props.expense.id, expense);
                        props.history.push('/');
                    }}
                />
                <button
                    className="button button--warning"
                    onClick={(e) => {
                        openModal();
                    }}
                >
                    Remove Expense
                </button>
                <RemoveModal
                    isOpen = {modalIsOpen}
                    handleModalClose = {closeModal}
                    handleRemoveExpense={handleRemoveExpense}
                />
            </div>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (id, expense) => dispatch(startEditExpense(id, expense)),
    onRemove: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);