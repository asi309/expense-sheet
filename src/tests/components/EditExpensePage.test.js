import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let onSubmit, onRemove, history, wrapper;

beforeEach(() => {
    onSubmit = jest.fn();
    history = {push: jest.fn()};
    onRemove = jest.fn();
    wrapper = shallow(
        <EditExpensePage
            onSubmit={onSubmit}
            onRemove={onRemove}
            history={history}
            expense={expenses[1]}
        />
    );
});

test('Should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('Should handle onRemove', () => {
    wrapper.find('button').simulate('Click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onRemove).toHaveBeenLastCalledWith(expenses[1].id);
});