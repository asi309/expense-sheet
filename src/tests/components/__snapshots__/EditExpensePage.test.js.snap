// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <h3>
    Edit Expense
  </h3>
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": 2,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
